codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -317,125 +314,60 @@ private Http2Stream requireStream(int streamId) {
-                        // writes just use empty buffer to avoid over-releasing. Have to use an empty buffer
-                        // as we may continue to write padding in subsequent frames.
-                        toWrite = data;
-                        data = Unpooled.EMPTY_BUFFER;
-                    }
-                    int writeablePadding = Math.min(allowedFrameSize - writeableData, padding);
-                    padding -= writeablePadding;
-                    bytesWritten += writeableData + writeablePadding;
-                    ChannelPromise writePromise;
-                    if (size == bytesWritten && !promise.isVoid()) {
-                        // Can use the original promise if it's the last write
-                        writePromise = promise;
-                    } else {
-                        // Create a new promise and listen to it for failure
-                        writePromise = ctx.newPromise();
-                        writePromise.addListener(this);
-                    }
-                    if (toWrite instanceof SlicedByteBuf && data instanceof CompositeByteBuf) {
-                        // If we're writing a subset of a composite buffer then we want to release
-                        // any underlying buffers that have been consumed. CompositeByteBuf only releases
-                        // underlying buffers on write if all of its data has been consumed and its refCnt becomes
-                        // 0.
-                        final CompositeByteBuf toFree = (CompositeByteBuf) data;
-                        writePromise.addListener(new ChannelFutureListener() {
-                            @Override
-                            public void operationComplete(ChannelFuture future) throws Exception {
-                                toFree.discardReadComponents();
-                            }
-                        });
-                    }
-                    frameWriter().writeData(ctx, stream.id(), toWrite, writeablePadding,
-                            size == bytesWritten && endOfStream, writePromise);
-                } while (size != bytesWritten && allowedBytes > bytesWritten);
-            } finally {
-                size -= bytesWritten;
-            }
+            int maxFrameSize = frameWriter().configuration().frameSizePolicy().maxFrameSize();
+            int bytesWritten = 0;
+            do {
+                int allowedFrameSize = Math.min(maxFrameSize, allowedBytes - bytesWritten);
+                int writeableData = Math.min(queue.readableBytes(), allowedFrameSize);
+                ChannelPromise writePromise = ctx.newPromise();
+                writePromise.addListener(this);
+                ByteBuf toWrite = queue.remove(writeableData, writePromise);
+
+                int writeablePadding = Math.min(allowedFrameSize - writeableData, padding);
+                padding -= writeablePadding;
+                bytesWritten += writeableData + writeablePadding;
```

Comment:
Instead of introducing `bytesWritten` and keeping track of it separately...could we just decrement `allowedBytes`?  This should save an extra variable and some extra arithmetic, and we could check against a constant `0` in the loop termination condition.